### ISSUE TITLE
Content-loop watchdog false-positives on markdown table generation, silently reported to clients as token-limit exhaustion

### ISSUE THREAD

**Issue by VU3RAZ** [Avarok-Cybersecurity/atlas#393]
**Environment:**
- Model: `Qwen/Qwen3.6-35B-A3B-FP8` (MTP, FP8-blockscaled)
- Atlas via `sparkrun run @atlas/qwen3.6-35b-a3b-fp8-mtp`
- Serve flags: `--max-seq-len 131072 --max-thinking-budget 4096 --enable-prefix-caching --speculative` (full command logged as: `spark serve Qwen/Qwen3.6-35B-A3B-FP8 --port 8888 --host 0.0.0.0 --gpu-memory-utilization 0.88 --max-seq-len 131072 --kv-cache-dtype bf16 --max-batch-size 2 --kv-high-precision-layers auto --scheduling-policy slai --max-prefill-tokens 32768 --oom-guard-mb 1024 --ssm-cache-slots 256 --mtp-quantization bf16 --max-thinking-budget 4096 --enable-prefix-caching --speculative`)

**Bug:**
The content-loop watchdog (`content-loop watchdog ENABLED (period-2…64 repetition detector)`) fires on legitimate markdown table output, mistaking the repeated border/separator characters (`│`, `─`, `┼`) for a genuine repetition loop. Generation is forcibly cut off immediately after the table header/border, before any data rows are emitted.

**Reproduction:**
Ask the model (via an OpenAI-compatible chat completion, tools active) to present tabular data, e.g. a list of scanned network hosts as a markdown table. Observed log sequence:
```
INFO spark::api::chat: Request: ... tools_active=true ...
INFO spark::scheduler::prefill_a_step: Thinking enabled, budget=Some(4096)
INFO spark::scheduler::emit_step: Grammar constrained decoding active: parser=qwen3_coder, tools=4
```
followed by the stream terminating right after the table header row, with no completion/`Done:` log line matching earlier successful (non-table) requests. Client (OpenAI-compatible consumer) receives a stream that appears to have hit a length limit, despite `max_tokens=65536` and `tool_max_tokens` far from exhausted.

**Impact:** Client applications (e.g. agentic coding tools) cannot distinguish this from genuinely hitting the token budget, since the finish reason presented is indistinguishable from `length`. There's no documented CLI flag or recipe config key to disable or tune the watchdog's sensitivity/period range — checked `sparkrun run --help`, the Atlas recipe schema, and the `atlas`/`atlas-recipes` repos and found no exposed control.

**Request:**
1. Exclude table-border/separator character sequences from the repetition heuristic, or raise the period range / add an allowlist for common formatting patterns.
2. Expose a CLI flag (e.g. `--content-loop-watchdog=off` or `--repetition-detector-period=<min>,<max>`) so users can disable/tune this per-deployment.
3. Report a distinct finish reason (e.g. `content_filter` or a custom reason) instead of overloading `length`, so clients don't misattribute this to token-limit exhaustion.

**Comment by VU3RAZ** [comment]
## Minimal, pi-independent reproduction

Confirmed this is not specific to markdown tables or to the `pi` client — a plain bulleted/bolded list of IPs in a tool-response follow-up reliably triggers the same false-positive watchdog stop, using a static curl payload with no live network calls.

**Repro payload** (`tool_test.json`):
```json
{
  "model": "MODEL_ID",
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "bash",
        "description": "Run a shell command",
        "parameters": {
          "type": "object",
          "properties": {
            "command": {"type": "string"},
            "timeout": {"type": "number"}
          },
          "required": ["command"]
        }
      }
    }
  ],
  "messages": [
    {"role": "user", "content": "Scan 10.10.10.0/24 and show me the results."},
    {"role": "assistant", "content": null, "tool_calls": [
      {"id": "call_1", "type": "function", "function": {"name": "bash", "arguments": "{\"command\": \"nmap -sn 10.10.10.0/24\"}"}}
    ]},
    {"role": "tool", "tool_call_id": "call_1", "content": "Nmap scan report for 10.10.10.1 (router.lan)\nNmap scan report for 10.10.10.5 (nas.lan)\nNmap scan report for 10.10.10.11 (printer.lan)\nNmap scan report for 10.10.10.21 (gb10.lan)\nNmap scan report for 10.10.10.34 (laptop.lan)"}
  ],
  "max_tokens": 8192,
  "enable_thinking": true
}
```

```bash
curl -s http://<host>:8888/v1/chat/completions -H "Content-Type: application/json" --data-binary @tool_test.json
```

**Result (run 1):**
```json
"content":"Here are the results of the scan for the 10.10.10.0/24 subnet:\n\n**Live Hosts Found:**\n*   **10.10.10.1**"
"finish_reason":"length"
"completion_tokens":217
```
Cut off after a single bullet item — nowhere near the 8192 `max_tokens` budget.

**Server log (run 1):**
```
WARN spark::scheduler::decode_logits_content: Content-loop watchdog fired (period-2…64 repeat); ending response early (rollback declined). content_tokens=48 output_len=216 matched_period=3 matched_repeats=3 reason=NoSsmSnapshot
INFO spark::scheduler::lifecycle: Done: 217 tokens (length) 59.0 tok/s, TTFT=490.5ms
```

**Server log (run 2, identical payload re-sent):**
```
WARN spark::scheduler::decode_logits_content: Content-loop watchdog fired (period-2…64 repeat); ending response early (rollback declined). content_tokens=64 output_len=242 matched_period=3 matched_repeats=3 reason=NoSsmSnapshot
INFO spark::scheduler::lifecycle: Done: 243 tokens (length) 58.1 tok/s, TTFT=57.3ms
```

Both runs: `matched_period=3, matched_repeats=3`. This confirms the watchdog is matching on the repeated `*   **10.10.10.X**` bullet/bold markdown pattern itself (period-3 token cycle from the bold-marker + bullet formatting), not genuine content repetition — the actual IP addresses differ each time it would continue, so there's no real semantic loop, just a syntactic one from consistent list formatting.

This also confirms the bug is **not specific to `pi`** or any particular client — it's fully reproducible via direct API calls with a static payload, no live network/nmap execution required.

**Comment by VU3RAZ** [comment]
## Confirmed: not FP8/quantization-specific, and fires from a second, independent code path

Re-tested this entire scenario against the Atlas-native NVFP4 recipe (`@atlas/qwen3.6-35b-a3b-nvfp4`, weights=NVFP4, still `--kv-cache-dtype fp8`) after ruling out a stale-container confound from an earlier restart. Same failure reproduces under NVFP4 weights — this is not specific to FP8 weight quantization.

More importantly, this NVFP4 run surfaced a watchdog firing from a **different code path** than every prior instance in this thread:

```
WARN spark::scheduler::emit_step: Content-loop watchdog fired in MTP/emit path (period-2…64 repeat); ending response content_tokens=96 output_len=97
INFO spark::scheduler::lifecycle: Done: 97 tokens (length) 101.4 tok/s, TTFT=502.1ms
```

vs. every earlier firing in this thread coming from `spark::scheduler::decode_logits_content`. This fired while the model was mid-way through listing plain comma-separated IP addresses (following a nmap-derived tool response), again well short of the token budget (97 tokens vs 8192 available).

The NVFP4 recipe runs with MTP speculative decoding enabled (`MTP gate: ... => ENABLED`), and this appears to be a second, independent trigger point for the same repetition heuristic, applied separately in the speculative-decode verify/emit path (`emit_step`) vs. the main decode path (`decode_logits_content`).

**Summary of everything confirmed in this thread:**
- Watchdog false-fires on markdown tables (period=3 on box-drawing chars)
- Watchdog false-fires on plain comma-separated IP lists, no markdown at all (period=0, `matched_repeats=0`, tagged `StreamUnsafe`)
- Reproduces identically under both FP8 and NVFP4 weight quantization — not a weight-precision issue
- Reproduces from two separate code locations (`decode_logits_content` and `emit_step`/MTP path)
- No CLI flag, recipe config key, or client-side prompt engineering (we tried instructing the model to avoid tables/lists via AGENTS.md) fully avoids it — only the markdown-specific trigger was avoidable that way; the plain-list trigger persists regardless

This looks like the repetition-detection heuristic itself (period-2…64 matching) is fundamentally too sensitive for any output containing structurally-similar repeated segments (list items, IPs, table rows), independent of quantization scheme or decode path. Given it's implemented in at least two separate places in the codebase, a fix likely needs to happen at the shared heuristic/threshold level rather than in either call site individually.

**Comment by EmmanuelNoscendi** [comment]
**Additional finding: the streaming-side line-repeat guard (`api/stream_guards`) has the same false-positive problem, and unlike the scheduler watchdog it cannot be disabled at all.**

Setup: `avarok/atlas-gb10:dev` on DGX Spark (GB10), Qwen3.6-35B-A3B-NVFP4, temp 0.

Repro: ask the model to copy 20 identical ~20-char lines, `stream: true` → the stream is killed after ~28 tokens with `finish_reason: "length"` and the already-buffered repeated content is withheld (0 lines delivered). Log: `spark::api::stream_guards: loop watchdog fired — repeated line (fuzzy-match) … occurrences=4`.
Any legitimate output with ≥4 similar lines in the 10 KB window (markdown tables, numbered lists, JSON arrays) trips it. Real-world hit: a 1381-token response (natural `stop` non-streaming) is cut at 652 tokens when streamed.

The asymmetry that makes this issue worse than the scheduler watchdog:
- `scheduler/emit_step.rs` checks `sched.levers.loop_watchdog()` and honors `ATLAS_DISABLE_WATCHDOGS=1` — after setting it, our repro passes non-streaming (natural stop).
- `api/chat_stream/handle_token.rs` calls `check_loop_watchdog(...)` (and the SimHash semantic guard) **unconditionally** — no lever, no env, no `MODEL.toml [behavior]` key, no TUI `/watchdog off` effect. Verified on current `main`.

Suggestion: gate the stream-side guards behind the same `SchedLevers::loop_watchdog` / `ATLAS_DISABLE_WATCHDOGS` controls (their placement in `stream_guards.rs` looks like a leftover from the `api/failures/` restore mentioned in the module comment). Happy to provide additional information.

**Comment by VU3RAZ** [comment]
No workaround on our end for the `stream_guards` path specifically — we hit the same family of false-positives on the scheduler-side watchdogs (`decode_logits_content` and the MTP `emit_step` path, see comments above) and never found a working override for those either (tried `ATLAS_DISABLE_WATCHDOGS`, sampling tuning, `AGENTS.md` prompt-side avoidance, NVFP4 vs FP8 — none fully avoided it). Ended up migrating this workload off Atlas to llama.cpp + native MTP speculative decoding instead, so we haven't tested against current `main`'s stream path. Wanted to flag in case it's a useful data point for prioritization — this looks like it makes streaming unusable for any structured output regardless of the scheduler-side levers.
